core/jvmMain/src/kotlinx/serialization/internal/Platform.kt

```diff
@@ -64,6 +58,36 @@ internal fun <T: Any> Class<T>.constructSerializerForGivenTypeArgs(vararg args:
     }
 }
 
+@Suppress("UNCHECKED_CAST")
+private fun <T: Any> Class<T>.findInNamedCompanion(vararg args: KSerializer<Any?>): KSerializer<T>? {
+    val namedCompanion = findNamedCompanionByAnnotation()
+    if (namedCompanion != null) {
+        invokeSerializerOnCompanion<T>(namedCompanion, *args)?.let { return it }
+    }
+
+    // fallback strategy for old compiler - try to locate plugin-generated singleton (without type parameters) serializer
+    return try {
+        declaredClasses.singleOrNull { it.simpleName == ("\$serializer") }
+            ?.getField("INSTANCE")?.get(null) as? KSerializer<T>
+    } catch (e: NoSuchFieldException) {
+        null
+    }
+}
+
+private fun <T: Any> Class<T>.findNamedCompanionByAnnotation(): Any? {
+    val companionClass = declaredClasses.firstOrNull { clazz ->
+        clazz.getAnnotation(NamedCompanion::class.java) != null
+    } ?: return null
+
+    return try {
+        val companion = getDeclaredField(companionClass.simpleName)
```

Comment:
I think this can be extracted and unified with `private fun Class<*>.companionOrNull`